src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,395 @@
+  RAY_CHECK_OK(actor_info_accessor_.AsyncUpdate(actor->GetActorID(), actor_table_data,
+                                                [this, actor](Status status) {
+                                                  RAY_CHECK_OK(status);
+                                                  // There is no promise that the node the
+                                                  // actor tied to is till alive as the
+                                                  // flush is asynchronously, so just
+                                                  // invoke `Schedule` which will lease
+                                                  // worker directly if the node is still
+                                                  // available or select a new one if not.
+                                                  Schedule(actor);
+                                                }));
+}
+
+std::vector<ActorID> GcsActorScheduler::CancelOnNode(const ClientID &node_id) {
+  // Remove all the actors from the map associated with this node, and return them as they
+  // will be reconstructed later.
+  std::vector<ActorID> actor_ids;
+
+  // Remove all actors in phase of leasing.
+  {
+    auto iter = node_to_actors_when_leasing_.find(node_id);
+    if (iter != node_to_actors_when_leasing_.end()) {
+      actor_ids.insert(actor_ids.end(), iter->second.begin(), iter->second.end());
+      node_to_actors_when_leasing_.erase(iter);
+    }
+  }
+
+  // Remove all actors in phase of creating.
+  {
+    auto iter = node_to_workers_when_creating_.find(node_id);
+    if (iter != node_to_workers_when_creating_.end()) {
+      for (auto &entry : iter->second) {
+        actor_ids.emplace_back(entry.second->TakeAssignedActorID());
+      }
+      node_to_workers_when_creating_.erase(iter);
+    }
+  }
+
+  // Remove the related remote lease client from remote_lease_clients_.
+  remote_lease_clients_.erase(node_id);
+  return actor_ids;
+}
+
+ActorID GcsActorScheduler::CancelOnWorker(const ClientID &node_id,
+                                          const WorkerID &worker_id) {
+  // Remove the worker which is creating actor and return the related actor.
+  return RemoveWorkerInPhaseOfCreating(node_id, worker_id);
+}
+
+void GcsActorScheduler::LeaseWorkerFromNode(std::shared_ptr<GcsActor> actor,
+                                            std::shared_ptr<rpc::GcsNodeInfo> node) {
+  RAY_CHECK(actor && node);
+
+  auto node_id = ClientID::FromBinary(node->node_id());
+  RAY_LOG(DEBUG) << "Start leasing worker from node " << node_id << " for actor "
+                 << actor->GetActorID();
+
+  auto retry_leasing_worker = [this, node_id, actor] {
+    // Reset the client so that we can reconnect when retry.
+    remote_lease_clients_.erase(node_id);
+    execute_after(io_context_,
+                  [this, node_id, actor] {
+                    if (auto node = gcs_node_manager_.GetNode(node_id)) {
```

Comment:
> This check is to make sure that the node is still alive, right? Can you add a comment explaining why this is necessary? Also, it would be good to add a comment explaining why it's okay to do nothing if the node is now dead.

Yes, it is just make sure that the node is still alive. If the node dead, maybe all actors in phase of leasing will be reconstructed before retry. So we need do nothing if the node is not alive in the async callback.

By the way, I totally agree with you that if the actor is not in the leasing map then do nothing. 🙂